### PR TITLE
Fix support for LOAD_FAST_AND_CLEAR

### LINKED ISSUE #9300: [Python 3.12] `LOAD_FAST_AND_CLEAR`
Python 3.12 introduced the `LOAD_FAST_AND_CLEAR` opcode. This turned out to be more involved than anticipated and it has been decided it will need attention outside of #9246 . This ticket tracks that effort.

### PR DESCRIPTION
**Pull request by sklam**
Fix #9300

### CODE REVIEW

File: numba/core/types/__init__.py [hunk 1]
```diff
@@ -28,6 +28,9 @@
 npy_rng = NumPyRandomGeneratorType('rng')
 npy_bitgen = NumPyRandomBitGeneratorType('bitgen')
 
+# _undef_var is used to represent undefined variables in the type system.
+_undef_var = UndefVar('_undef_var') #
```
Comment (stuartarchibald) [inline numba/core/types/__init__.py:32]:
```suggestion
_undef_var = UndefVar('_undef_var')
```

File: numba/core/typeinfer.py [hunk 2]
```diff
@@ -53,6 +53,11 @@ def __init__(self, context, var):
 
     def add_type(self, tp, loc):
         assert isinstance(tp, types.Type), type(tp)
+        # Special case for _undef_var.
+        # If the typevar is a _undef_var, use the incoming type directly.
```
Comment (stuartarchibald) [inline numba/core/typeinfer.py:57]:
```suggestion
        # If the typevar is the _undef_var, use the incoming type directly.
```
Presumably there's only one `types._undef_var` as it's on the `types.__init__`?

File: numba/cpython/builtins.py [hunk 3]
```diff
@@ -1003,3 +1003,11 @@ def impl(object=""):
             return repr(object)
     return impl
 
+
+@lower_cast(types.UndefVar, types.Any)
+def cast_undef_var_to_any(context, builder, fromty, toty, val):
+    """UndefVar cast to anything.
```
Comment (stuartarchibald) [inline numba/cpython/builtins.py:1009]:
```suggestion
    """UndefVar can cast to anything.
```

File: numba/tests/test_typeinfer.py [hunk 4]
```diff
@@ -707,6 +708,25 @@ def confuse_typer(x):
...
         self.assertTrue(all(vt == types.float64 for vt in return_vars.values()))
 
+    @unittest.skipUnless("LOAD_FAST_AND_CLEAR" in dis.opmap,
+                         "Requires LOAD_FAST_AND_CLEAR opcode")
+    def test_load_fast_and_clear(self):
+        @njit
+        def foo(a):
+            [x for x in (0,)]
+            if a:
+                # can't be a constant due to constant propagation issues.
```
Comment (stuartarchibald) [inline numba/tests/test_typeinfer.py:718]:
```suggestion
                # the test code cannot use a constant here due to constant
                # propagation issues.
```

File: numba/tests/test_typeinfer.py [hunk 5]
```diff
@@ -707,6 +708,25 @@ def confuse_typer(x):
...
+                x = 3 + a
+            x += 10
+            return x
+
+        self.assertEqual(foo(True), foo.py_func(True))
+        # Interpreted version should raise an exception
+        with self.assertRaises(UnboundLocalError):
+            foo.py_func(False)
+        # Compiled version returns 10 as x is zero initialized.
+        self.assertEqual(foo(False), 10)
```
Comment (stuartarchibald) [inline numba/tests/test_typeinfer.py:728]: Note: This is indeed the case, it's unusual but valid.

File: numba/core/types/misc.py [hunk 6]
```diff
@@ -33,6 +33,18 @@ def is_precise(self):
...
+    A type that is created by Expr.undef to represent an undefined variable.
+    This type can be promoted to any other type.
+    This is introduced to handle Python 3.12 LOAD_FAST_AND_CLEAR.
+    """
+
+    def can_convert_to(self, typingctx, other):
+        return Conversion.promote
+
+
+
```
Comment (stuartarchibald) [inline numba/core/types/misc.py:47]:
```suggestion
```
2 empty lines between class definitions.

File: numba/core/lowering.py [hunk 7]
```diff
@@ -1413,6 +1413,11 @@ def lower_expr(self, resty, expr):
         elif expr.op == 'null':
             return self.context.get_constant_null(resty)
 
+        elif expr.op == 'undef':
+            # Numba do not raise UnboundLocalError for undefined variables.
```
Comment (stuartarchibald) [inline numba/core/lowering.py:1417]:
```suggestion
            # Numba does not raise an UnboundLocalError for undefined variables.
```

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the patch, this seems to fix most of the issues associated with the `LOAD_FAST_AND_CLEAR` op code. In testing this, the following sample is a variant on a test from the CPython test suite, specifically: https://github.com/python/cpython/blob/d4a6229afe10d7e5a9a59bf9472f36d7698988db/Lib/test/test_listcomps.py#L368
```python
@njit
def foo():
    if False:
        x = 0
    [x for x in (1,)]
    return x

foo()
```
which results in a `TypeError: cannot convert native _undef_var to Python object`.

**Comment by sklam** [comment]
RE: https://github.com/numba/numba/pull/9315#pullrequestreview-1755908955

addressed in https://github.com/numba/numba/pull/9315/commits/3f5ebbb5ba3857f79acf87f17ab6648ca592d13c

the type inferer will scan for use of `_undef_var` in call arguments and in return types.

**Comment by sklam** [comment]
CI is failing due to:

- ambiguous cast between undef-to-any and optional cast
- objectmode not handling Expr.undef

**Review by esc — APPROVED** [review]
LGTM

**Pull request merged**